### ISSUE TITLE
Problem with caching and dependent modules

### ISSUE THREAD

**Issue by mholo65** [statiqdev/Statiq.Framework#75]
This is rather problematic to explain, so instead, I'll throw some code to reproduce here:

```csharp
using Statiq.App;
using Statiq.Common;
using Statiq.Core;
using System.Linq;
using System.Threading.Tasks;

namespace ConsoleApp3
{
    class Program
    {
        static Task Main(string[] args)
        {
            return Bootstrapper
                .CreateDefault(args)
                .RunAsync();
        }
    }

    public class Pipeline1 : Pipeline
    {
        public Pipeline1()
        {
            InputModules = new ModuleList { new ReadFiles("1.txt") };

            ProcessModules = new ModuleList
            {
                new CacheDocuments(
                    new SetContent(Config.FromDocument(async x => $"Modified {await x.GetContentStringAsync()}"))
                    )
            };
        }
    }

    public class Pipeline2 : Pipeline
    {
        public Pipeline2()
        {
            InputModules = new ModuleList { new ReadFiles("2.txt") };

            ProcessModules = new ModuleList
            {
                new CacheDocuments(
                    new SetContent(Config.FromDocument(async x => $"Modified {await x.GetContentStringAsync()}"))
                    )
            };
        }
    }

    public class Pipeline3 : Pipeline
    {
        public Pipeline3()
        {
            Dependencies.AddRange(new[]
            {
                nameof(Pipeline1),
                nameof(Pipeline2)
            });

            InputModules = new ModuleList { new ReadFiles("3.txt") };

            ProcessModules = new ModuleList
            {
                new CacheDocuments(
                    new SetMetadata("foo", Config.FromValue("foo"))
                    )
            };

            TransformModules = new ModuleList
            {
                new ConcatDocuments(
                    nameof(Pipeline1),
                    nameof(Pipeline2)),
                new CacheDocuments(
                    new SetContent(
                        Config.FromDocument(async (doc, ctx) =>
                            await doc.GetContentStringAsync() + ctx.Inputs.Single(x => x.ContainsKey("foo")).GetString("foo")))
                    )
            };

            OutputModules = new ModuleList
            {
                new WriteFiles()
            };
        }
    }
}

```

Above, we have three pipelines; `Pipeline1`, `Pipeline2` and `Pipeline3`. `Pipeline1` and `Pipeline2` are identical, but operate on different set of files. `Pipeline3` operates on a third set of file, but is also dependent on `Pipeline1` and `Pipeline2` and includes their documents in the `Transform` phase. `Pipeline3` is also responsible for writing all the files to disk.

So what's happening then when running these in `preview`mode?

On first run, everything is ok:

```
| Pipeline  | Input      | Process   | Transform | Output    | Total Time |
|-------------------------------------------------------------------------|
| Pipeline1 | 1 (111 ms) | 1 (90 ms) |           |           | 201 ms     |
| Pipeline2 | 1 (111 ms) | 1 (90 ms) |           |           | 201 ms     |
| Pipeline3 | 1 (111 ms) | 1 (10 ms) | 3 (18 ms) | 3 (26 ms) | 165 ms     |
```

On second run, when I modify document `3.txt` (which is read by `Pipeline3`), everything also works fine. (From the process times, we can also see that caching is working really well).
```
| Pipeline  | Input     | Process  | Transform | Output   | Total Time |
|----------------------------------------------------------------------|
| Pipeline1 | 1 (54 ms) | 1 (5 ms) |           |          | 59 ms      |
| Pipeline2 | 1 (54 ms) | 1 (6 ms) |           |          | 60 ms      |
| Pipeline3 | 1 (43 ms) | 1 (3 ms) | 3 (5 ms)  | 3 (8 ms) | 59 ms      |
```

But then, if I edit document `1.txt` or `2.txt` (which are documents that `Pipeline1` and `Pipeline2` are operates on), then `ctx.Inputs.Single(x => x.ContainsKey("foo"))` in `Pipeline3` `Transform` phase will throw an exception saying that sequence contains no matching elements. (also not the weird `2 files have changed`, even though I only modified one).
```
[INFO] 2 files have changed, re-executing
[INFO] Executing 3 pipelines (Pipeline1, Pipeline2, Pipeline3)
[INFO] Cleaned temp directory: temp
[INFO] Cleaned output directory: output
[INFO] -> Pipeline2/Input » Starting Pipeline2 Input phase execution... (0 input document(s), 1 module(s))
[INFO] -> Pipeline1/Input » Starting Pipeline1 Input phase execution... (0 input document(s), 1 module(s))
[INFO] -> Pipeline3/Input » Starting Pipeline3 Input phase execution... (0 input document(s), 1 module(s))
[INFO]    Pipeline3/Input » Finished Pipeline3 Input phase execution (1 output document(s), 12 ms)
[INFO]    Pipeline2/Input » Finished Pipeline2 Input phase execution (1 output document(s), 12 ms)
[INFO]    Pipeline1/Input » Finished Pipeline1 Input phase execution (1 output document(s), 13 ms)
[INFO] -> Pipeline1/Process » Starting Pipeline1 Process phase execution... (1 input document(s), 1 module(s))
[INFO] -> Pipeline2/Process » Starting Pipeline2 Process phase execution... (1 input document(s), 1 module(s))
[INFO]    Pipeline2/Process » Finished Pipeline2 Process phase execution (1 output document(s), 8 ms)
[INFO]    Pipeline1/Process » Finished Pipeline1 Process phase execution (1 output document(s), 16 ms)
[INFO] -> Pipeline3/Process » Starting Pipeline3 Process phase execution... (1 input document(s), 1 module(s))
[INFO]    Pipeline3/Process » Finished Pipeline3 Process phase execution (1 output document(s), 3 ms)
[INFO] -> Pipeline3/Transform » Starting Pipeline3 Transform phase execution... (1 input document(s), 2 module(s))
[ERRO] Pipeline3/Transform » CacheDocuments » Error while executing module SetContent in Pipeline3/Transform: Sequence contains no elements
[ERRO] Error while executing module CacheDocuments in Pipeline3/Transform: Sequence contains no elements
```

**UPDATE**
If I change the `Single()` to a `FirstOrDefault()` and return an empty string if null, like this:
```csharp
ctx.Inputs.FirstOrDefault(x => x.ContainsKey("foo"))?.GetString("foo") ?? string.Empty
```
I will of course not get an exception, and summary will show correct amount of documents:
```
| Pipeline  | Input     | Process   | Transform | Output    | Total Time |
|------------------------------------------------------------------------|
| Pipeline1 | 1 (10 ms) | 1 (7 ms)  |           |           | 17 ms      |
| Pipeline2 | 1 (10 ms) | 1 (13 ms) |           |           | 23 ms      |
| Pipeline3 | 1 (10 ms) | 1 (3 ms)  | 3 (10 ms) | 3 (10 ms) | 33 ms      |
```
But if I inspect the content of the changed file (`2.txt` in this case) I can of course see that `foo` has not been appended to that document in the output.

Am I doing something wrong, is this supposed to work? Wouldn't mind digging deeper and fix this if I'd get some pointers on where to start looking for a fix.

//cc @daveaglick

**Comment by daveaglick** [comment]
Awesome clean repro, thanks! Still trying to wrap my brain around what should be happening here, but my initial thought is that it's a bug (as opposed to incorrect use).

The case where `file3.txt` is changed seems to be working as designed. The `file1.txt` and `file2.txt` documents are (hopefully) being output by the cache and then fed to `Pipeline3`.

I think what might be happening is something with the way `CacheDocuments` determines if it should run and what to set as context inputs when it does. When either `file1.txt` or `file2.txt` are changed, those pipelines appear to run correctly and output the updated document, but because `file3.txt` which is the _only_ input to the `CacheDocuments` module in `Pipeline3` isn't changed, it'll have the same hash code, and `CacheDocuments` will kick into cache mode. That was probably happening in `Pipeline1` and `Pipeline2` as well when `file3.txt` was changed, but we didn't notice because those cached modules didn't try to access `IExecutionContext.Inputs`. Best guess now is that `Inputs` maybe isn't getting set correctly when in cache mode. I'm also not entirely sure why the `CacheDocuments` in `Pipeline3` is even executing it's child modules when the input `file3.txt` isn't changed (which is a whole other problem because something we'll want to invalidate the cache for dependent documents that aren't inputs).

**Comment by daveaglick** [comment]
Hmm...just noticed that `Pipeline3` has two `CacheDocuments` and it's crashing in the second on transform - I still think the problem is related to what I wrote above, but might be more subtle than I first thought. Great bug find.

**Comment by daveaglick** [comment]
Pretty sure this is the problem:
![2020-01-25_14h37_15](https://user-images.githubusercontent.com/1020407/73126588-a2550280-3f82-11ea-8148-a5a8ef9d98e9.png)

The `CacheDocuments` module only sends missed documents as inputs to the child modules. That makes sense because otherwise we'd end up processing the hits (which we don't want to do because what would be the point of the cache then). On the other hand, it clearly messes up situations like this.

Not sure about a good resolution - thoughts @mholo65?
- Could leave the behavior like it is but improve error message by throwing or something any time `Inputs` is accessed from the child module of a cache - "cached modules should not access inputs"
- Could leave the behavior like it is but send documents through the child modules one at a time to reinforce that caching is only for one-to-one situations - `Inputs` would only ever contain a single (current) document in this can and would need to document that well
- Could consider changing the behavior and somehow setting `Inputs` to _all_ documents, hit or miss, but only running child modules for the missed documents (this would require a little plumbing work since `Inputs` and the documents being processed are always the same right now).

**Comment by mholo65** [comment]
My current solution is to cache the wanted input document (handlebars layout in my case) in the module. I actually find it from Inputs in BeforeExecution. That works for now.

IMO, best solution is the last option, as you don't necessary know how a module operates (for example a render module pulling layout from Inputs). Current approach enforces you to take CacheDocuments module behaviour into account when writing a module, which is not ideal as CacheDocuments should be treated as a decorator.

That's my two cents. Lot's of plumbing, but maybe worth it in the end. A warning or guidelines also works, but please don't throw, as you can read from Inputs the first time, and then persist it inside the module.

**Comment by daveaglick** [comment]
This turns out to be a tough one...

I agree in theory that `CacheDocuments` should be invisible to child modules, but on the other hand, I'm not sure it's possible. Here's the problem:
- We have all the input documents
- We have hash codes for each so we can figure out cache hits and misses
- There's already an assumption that child modules will act on a one-to-one basis from input document to output document, otherwise we wouldn't be able to partially cache and would have to rerun _all_ documents when _any_ document changed
- Modules use `IExecutionContext.Inputs` as the thing they operate on - there isn't any other concept of "work on this particular document" because the module is free to perform parallel operations, aggregate, etc. the inputs however is appropriate - this simple interface is the basis for everything else
- Which leaves us with...if the child modules of `CacheDocuments` get the full set of inputs, **there's no way to indicate which inputs should be "run" and which should be ignored** (ignored because they were cache hits and we already know the corresponding output)
  - Further complicating that is even if we could "ignore" certain inputs, we'd have to remove them at the end and replace with the cache hits

So we could enforce the one-to-one behavior further by essentially treating `CacheDocuments` like `ForEachDocument` and executing the child modules once for each individual document instead of once for all missed documents. That would mean child modules couldn't do their work concurrently across documents, but would also make the semantics a bit cleaner (maybe). I'm not even sure it's a good idea - as long as it's clear modules shouldn't really use _data_ from other inputs (but can use the whole input collection to optimize processing though concurrency, etc.), we're better off leaving it as-is by passing just the documents that still need processing (the misses).

TL;DR: the current behavior might actually be the most correct.

The only other thing I can think to do is add something like `IExecutionContext.SomeOtherProperty` that contains a bigger set of documents. In almost every situation it would be the same as `Inputs` except inside the cache module where `Inputs` would be the cache misses and the other property (any ideas on a name?) would be the broader set of all documents. That could work, and maybe there are other cases were inputs != larger set of documents, though I can't think of any right now.

**Comment by daveaglick** [comment]
Actually, this might already be solved!

I started thinking about that new property in `IExecutionContext` should be called and what it should do and settled on maybe it should be called `ParentInputs` and contain the inputs to the parent module like `CacheDocuments` or `ForEachDocument` - that way any time a parent module filters or modifies documents to pass to it’s child modules the child modules could still get at the original set of inputs.

Turns out there’s already a `IExecutionContext.Parent` which contains the entire parent context (and on up the hierarchy). So from inside a module in `CacheDocuments` you can get to the set of all inputs through `IExecutionContext.Parent.Inputs`. If you access that to get your layouts (or in the case of the repro, to get the single), does it solve the problem?

There’s still the issue of making sure we correctly count missed due to dependent documents. For example, in the repro that third case is attempting to change a document that hadn’t been a cache miss because the original document didn’t change and the hash was the same. In other words, there’s currently nothing to say “this document is invalid if it’s hash changes _or_ if these other documents change”. Best way I can see to handle that case is add a setting to `CacheDocuments` that lets you specify conditions on other documents that if they change invalidates the entire cache - not sure how we’d do that on a per-document based though.

**Comment by daveaglick** [comment]
Slept on this and thought about it more today - and the more I think about it, the more I'm convinced the current behavior is correct and just needs to be better documented. I don't think there's a good way to make the `CacheDocuments` module transparent - by the nature of not wanting to reprocesses hits, it has to send different inputs. Hopefully between documenting that the inputs to child modules will only be misses and the full set of documents is available via `Parent.Inputs` most of the confusion will be addressed.

I've created #78 to track specifying more complex caching dependencies.

**Comment by daveaglick** [comment]
Updated the docs to make the behavior clearer and backed it up with tests. Going to call this one done.
